pkg/apiserver/api_installer.go

```diff
@@ -321,10 +321,12 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 
 			// list or post across namespace.
 			// TODO: more strongly type whether a resource allows these actions on "all namespaces" (bulk delete)
-			namer = scopeNaming{scope, a.group.Linker, gpath.Join(a.prefix, itemPath), true}
-			actions = appendIf(actions, action{"LIST", resource, params, namer}, isLister)
-			actions = appendIf(actions, action{"POST", resource, params, namer}, isCreater)
-			actions = appendIf(actions, action{"WATCHLIST", "watch/" + resource, params, namer}, allowWatchList)
+			if !hasSubresource {
+				namer = scopeNaming{scope, a.group.Linker, gpath.Join(a.prefix, itemPath), true}
+				actions = appendIf(actions, action{"LIST", resource, params, namer}, isLister)
+				actions = appendIf(actions, action{"POST", resource, params, namer}, isCreater)
```

Comment:
I think we need to define something in the rest handler that specifies whether it supports all-namespaces, and whether those verbs are allowed (LIST/POST).  As a short term fix this is acceptable, but we'll have to come back with a patch that supports this.

The way I would envision the long-term fix is:

pkg/api/rest/rest.go:

  type CrossNamespaceScoped interface {
    SupportsCrossNamespace() <returns something that identifies which interface methods among rest.Creator rest.Updater rest.Deletor allow cross namespace, maybe a rest.Storage object again>
  }

We want to stop assuming that every resource supports cross namespace calls (that's lead to several security holes), but we want to very explicitly tie the interface to the binding.

----- Original Message -----
> > @@ -321,10 +321,12 @@ func (a *APIInstaller) registerResourceHandlers(path
> > string, storage rest.Storag
> >  
> >  			// list or post across namespace.
> >  			// TODO: more strongly type whether a resource allows these actions on
> >  			"all namespaces" (bulk delete)
> > -			namer = scopeNaming{scope, a.group.Linker, gpath.Join(a.prefix,
> > itemPath), true}
> > -			actions = appendIf(actions, action{"LIST", resource, params, namer},
> > isLister)
> > -			actions = appendIf(actions, action{"POST", resource, params, namer},
> > isCreater)
> > -			actions = appendIf(actions, action{"WATCHLIST", "watch/" + resource,
> > params, namer}, allowWatchList)
> > +			if !hasSubresource {
> > +				namer = scopeNaming{scope, a.group.Linker, gpath.Join(a.prefix,
> > itemPath), true}
> > +				actions = appendIf(actions, action{"LIST", resource, params, namer},
> > isLister)
> > +				actions = appendIf(actions, action{"POST", resource, params, namer},
> > isCreater)
> 
> This is to stop adding another POST method for pods/ while processing
> pods/binding.
> It is correctly added while processing pods.
> Right now,  it is added twice resulting in the
> ```#/apis/56/operations/2/method: Operation method already defined: POST```
> error
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/9991/files#r32751166